File: specs/typespec/org/references.go
package org

import (
	"vetchium-api-server.typespec/common"
	hub "vetchium-api-server.typespec/hub"
)

type RequestReferencesRequest struct {
	CandidacyID      string                  `json:"candidacy_id"`
	MaxReferences    int32                   `json:"max_references"`
	ResponseDeadline string                  `json:"response_deadline"`
	Questions        []hub.ReferenceQuestion `json:"questions"`
}

type RequestReferencesResponse struct {
	RequestID string `json:"request_id"`
}

type RequestIdRequest struct {
	RequestID string `json:"request_id"`
}

type OrgReferenceNomination struct {
	NominationID       string                       `json:"nomination_id"`
	NomineeHandle      string                       `json:"nominee_handle"`
	NomineeDisplayName string                       `json:"nominee_display_name"`
	SharedDomain       string                       `json:"shared_domain"`
	OverlapStartYear   int32                        `json:"overlap_start_year"`
	OverlapEndYear     int32                        `json:"overlap_end_year"`
	State              hub.ReferenceNominationState `json:"state"`
	NominatedAt        string                       `json:"nominated_at"`
	SubmittedAt        *string                      `json:"submitted_at,omitempty"`
}

type ListReferenceNominationsResponse struct {
	Nominations []OrgReferenceNomination `json:"nominations"`
}

type ReferenceResponseAnswer struct {
	QuestionID   string `json:"question_id"`
	QuestionText string `json:"question_text"`
	ResponseText string `json:"response_text"`
}

type OrgReferenceResponse struct {
	NominationID       string                    `json:"nomination_id"`
	NomineeHandle      string                    `json:"nominee_handle"`
	NomineeDisplayName string                    `json:"nominee_display_name"`
	SharedDomain       string                    `json:"shared_domain"`
	OverlapStartYear   int32                     `json:"overlap_start_year"`
	OverlapEndYear     int32                     `json:"overlap_end_year"`
	Answers            []ReferenceResponseAnswer `json:"answers"`
	SubmittedAt        string                    `json:"submitted_at"`
}

type ListReferenceResponsesResponse struct {
	Responses           []OrgReferenceResponse   `json:"responses"`
	DeclinedNominations []OrgReferenceNomination `json:"declined_nominations"`
}

func (r RequestReferencesRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.CandidacyID == "" {
		errs = append(errs, common.ValidationError{Field: "candidacy_id", Message: "Must be a non-empty string"})
	}
	if r.MaxReferences < 1 || r.MaxReferences > 5 {
		errs = append(errs, common.ValidationError{Field: "max_references", Message: "Must be between 1 and 5"})
	}
	if r.ResponseDeadline == "" {
		errs = append(errs, common.ValidationError{Field: "response_deadline", Message: "Must be a non-empty string"})
	}
	if len(r.Questions) < 1 || len(r.Questions) > 10 {
		errs = append(errs, common.ValidationError{Field: "questions", Message: "Must have 1-10 items"})
	}
	for i, q := range r.Questions {
		if len(q.Text) < 10 || len(q.Text) > 500 {
			errs = append(errs, common.ValidationError{Field: "questions[" + string(rune(i)) + "].text", Message: "Must be between 10 and 500 characters"})
		}
		if q.MinChars < 0 {
			errs = append(errs, common.ValidationError{Field: "questions[" + string(rune(i)) + "].min_chars", Message: "Must be non-negative"})
		}
		if q.MaxChars < 1 || q.MaxChars > 4000 {
			errs = append(errs, common.ValidationError{Field: "questions[" + string(rune(i)) + "].max_chars", Message: "Must be between 1 and 4000"})
		}
	}
	return errs
}

func (r RequestIdRequest) Validate() []common.ValidationError {
	var errs []common.ValidationError
	if r.RequestID == "" {
		errs = append(errs, common.ValidationError{Field: "request_id", Message: "Must be a non-empty string"})
	}
	return errs
}

File: specs/typespec/org/references.ts
import type { Handle } from "../hub/hub-users.js";
import type {
	ReferenceQuestion,
	ReferenceNominationState,
} from "../hub/references.js";

export interface RequestReferencesRequest {
	candidacy_id: string;
	max_references: number;
	response_deadline: string;
	questions: ReferenceQuestion[];
}

export interface RequestReferencesResponse {
	request_id: string;
}

export interface RequestIdRequest {
	request_id: string;
}

export interface OrgReferenceNomination {
	nomination_id: string;
	nominee_handle: Handle;
	nominee_display_name: string;
	shared_domain: string;
	overlap_start_year: number;
	overlap_end_year: number;
	state: ReferenceNominationState;
	nominated_at: string;
	submitted_at?: string;
}

export interface ListReferenceNominationsResponse {
	nominations: OrgReferenceNomination[];
}

export interface OrgReferenceResponse {
	nomination_id: string;
	nominee_handle: Handle;
	nominee_display_name: string;
	shared_domain: string;
	overlap_start_year: number;
	overlap_end_year: number;
	answers: Array<{
		question_id: string;
		question_text: string;
		response_text: string;
	}>;
	submitted_at: string;
}

export interface ListReferenceResponsesResponse {
	responses: OrgReferenceResponse[];
	declined_nominations: OrgReferenceNomination[];
}

export interface ValidationError {
	field: string;
	message: string;
}

export function validateRequestReferencesRequest(
	req: unknown
): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.candidacy_id !== "string" || r.candidacy_id.trim() === "") {
		errors.push({
			field: "candidacy_id",
			message: "Must be a non-empty string",
		});
	}

	if (
		typeof r.max_references !== "number" ||
		r.max_references < 1 ||
		r.max_references > 5
	) {
		errors.push({
			field: "max_references",
			message: "Must be between 1 and 5",
		});
	}

	if (
		typeof r.response_deadline !== "string" ||
		r.response_deadline.trim() === ""
	) {
		errors.push({
			field: "response_deadline",
			message: "Must be a non-empty string",
		});
	}

	if (!Array.isArray(r.questions)) {
		errors.push({ field: "questions", message: "Must be an array" });
	} else {
		if (r.questions.length < 1 || r.questions.length > 10) {
			errors.push({
				field: "questions",
				message: "Must have 1-10 items",
			});
		}
		for (let i = 0; i < r.questions.length; i++) {
			const q = r.questions[i];
			if (!q || typeof q !== "object") {
				errors.push({
					field: `questions[${i}]`,
					message: "Must be an object",
				});
				continue;
			}
			const qObj = q as Record<string, unknown>;
			if (
				typeof qObj.text !== "string" ||
				qObj.text.length < 10 ||
				qObj.text.length > 500
			) {
				errors.push({
					field: `questions[${i}].text`,
					message: "Must be between 10 and 500 characters",
				});
			}
			if (typeof qObj.min_chars !== "number" || qObj.min_chars < 0) {
				errors.push({
					field: `questions[${i}].min_chars`,
					message: "Must be a non-negative number",
				});
			}
			if (
				typeof qObj.max_chars !== "number" ||
				qObj.max_chars < 1 ||
				qObj.max_chars > 4000
			) {
				errors.push({
					field: `questions[${i}].max_chars`,
					message: "Must be between 1 and 4000",
				});
			}
		}
	}

	return errors;
}

export function validateRequestIdRequest(req: unknown): ValidationError[] {
	const errors: ValidationError[] = [];
	if (!req || typeof req !== "object") {
		return [{ field: "$root", message: "Request body is required" }];
	}
	const r = req as Record<string, unknown>;

	if (typeof r.request_id !== "string" || r.request_id.trim() === "") {
		errors.push({ field: "request_id", message: "Must be a non-empty string" });
	}

	return errors;
}
